### PR TITLE
Fix gtest multithread crash

### PR DESCRIPTION
**Pull request by WeiChungChang**
This change allows the use of GoogleTest's ASSERT() and EXPECT() macros on multiple threads. The failures will be correctly reported and no longer crashes the app.

When GoogleTest is used in multi-thread mode (i.e. its macros are used on thread other than the main thread), it may crash as GetTestPartResultReporterForCurrentThread() returns nullptr on a forked thread.

This is due to that the ThreadLocal class in our GTEST_OS_STARBOARD implementation doesn't return the default value on newly created threads:
The ThreadLocal ctor accepts a default value of T to be returned on a thread where the thread local value hasn't been overridden. The existing implementation incorrectly returns T() on a new thread where the value has never been set. When T is a pointer, T() is nullptr and causes crash when dereferenced.

Now the default value to ThreadLocal ctor is stored and a copy of it will be returned on a new thread when the thread local value hasn't been set. This ensures that GetTestPartResultReporterForCurrentThread() returns the correct TestPartResultReporterInterface*, and failures of GoogleTest's ASSERT and EXPECT macros on a different thread will be correctly reported and no longer crashes the app.

b/230877781

### CODE REVIEW

File: testing/gtest/src/gtest.cc [hunk 1]
```diff
@@ -737,6 +737,10 @@ void UnitTestImpl::SetGlobalTestPartResultReporter(
 // Returns the test part result reporter for the current thread.
 TestPartResultReporterInterface*
 UnitTestImpl::GetTestPartResultReporterForCurrentThread() {
+  if (!per_thread_test_part_result_reporter_.get()) {
+    per_thread_test_part_result_reporter_.set(
```
[the remark this answers is not in the document]
Comment (WeiChungChang) [reply testing/gtest/src/gtest.cc]: Thanks a lot to the suggestion.

**Comment by xiaomings** [comment]
Consider turning the comment into:

This change allows the use of GoogleTest's ASSERT() and EXPECT() macros on multiple threads. The failures will be correctly reported and no longer crashes the app.

When GoogleTest is used in multi-thread mode (i.e. its macros are used on thread other than the main thread), it may crash as GetTestPartResultReporterForCurrentThread() returns nullptr on a forked thread.

This is due to that the ThreadLocal<T> class in our GTEST_OS_STARBOARD implementation doesn't return the default value on newly created threads:
The ThreadLocal ctor accepts a default value of T to be returned on a thread where the thread local value hasn't been overridden.  The existing implementation incorrectly returns T() on a new thread where the value has never been set.  When T is a pointer, T() is nullptr and causes crash when dereferenced.

Now the default value to ThreadLocal ctor is stored and a copy of it will be returned on a new thread when the thread local value hasn't been set.  This ensures that GetTestPartResultReporterForCurrentThread() returns the correct  `TestPartResultReporterInterface*`, and failures of GoogleTest's ASSERT and EXPECT macros on a different thread will be correctly reported and no longer crashes the app.

b/230877781

File: testing/gtest/include/gtest/internal/gtest-port.h [hunk 2]
```diff
@@ -1570,13 +1573,14 @@ class ThreadLocal {
...
     } else {
-      T* new_value = new T();
+      T* new_value = new T(value_);
       bool is_set = SbThreadSetLocalValue(key_, new_value);
       SB_CHECK(is_set);
       return new_value;
     }
   }
 
+  T value_;
```
Comment (xiaomings) [inline testing/gtest/include/gtest/internal/gtest-port.h:1583]: nit: default_value_

Comment (WeiChungChang) [reply testing/gtest/include/gtest/internal/gtest-port.h]: done, thanks

**Review by jasonzhangxx — APPROVED** [review]

**Pull request merged**